kyverno/kyverno
# fix: generate events for scanning VAPs in reports controller

## Explanation
This PR generates events for scanning validating admission policies in reports controller



## Related issue
Closes #8781 


## Milestone of this PR
/milestone 1.11.0


## What type of PR is this
/kind fix


## Proposed Changes



### Proof Manifests
1. Create a deployment with replicas greater than 3:
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx-deployment-2
  labels:
    app: nginx-2
spec:
  replicas: 7
  selector:
    matchLabels:
      app: nginx-2
  template:
    metadata:
      labels:
        app: nginx-2
    spec:
      containers:
      - name: nginx
        image: nginx:1.14.2
        ports:
        - containerPort: 80
```
2. Create a validating admission policy that checks the deployment replicas:
```
apiVersion: admissionregistration.k8s.io/v1alpha1
kind: ValidatingAdmissionPolicy
metadata:
  name: check-deployment-replicas
spec:
  matchConstraints:
    resourceRules:
    - apiGroups:   ["apps"]
      apiVersions: ["v1"]
      operations:  ["CREATE", "UPDATE"]
      resources:   ["deployments"]
  validations:
    - expression: "object.spec.replicas <= 3"
      message: "Deployment spec.replicas must be less than 3."
```
3. Check the events:
```
kubectl get events | grep validatingadmissionpolicy
28m         Warning   PolicyViolation     validatingadmissionpolicy/check-deployment-replicas   Deployment default/nginx-deployment-2: [check-deployment-replicas] fail; Deployment spec.replicas must be less than 3.
```
4. Check the reports controller logs:
```
I1030 13:44:04.546094       1 event_broadcaster.go:318] "Event occurred" object="check-deployment-replicas" kind="ValidatingAdmissionPolicy" apiVersion="admissionregistration.k8s.io/v1alpha1" type="Warning" reason="PolicyViolation" action="Resource Passed" note="Deployment default/nginx-deployment-2: [check-deployment-replicas] fail; Deployment spec.replicas must be less than 3."
```



## Checklist



  
  
  
  

## Further Comments
